### DISCUSSION TITLE
How should one use the Logger ? (best practices)

### DISCUSSION THREAD

**Discussion by JulienChaos** [log4cplus/log4cplus#633] [Q&A] [answered]
I would like to know the best practice for using the Logger for a library or an application. Can I share a static instance for my application to simplify the syntax and avoid storing a member variable?

```c++
// MyLogger.hpp

#include <log4cplus/loggingmacros.h>

extern log4cplus::Logger logger;

static void init();

// MyLogger.cpp

#include "MyLogger.hpp"

log4cplus::Logger logger

void init()
{
//init
  logger = Logger::getInstance("AppName");
}

// main.cpp

#include "MyLogger.hpp"

int maint(int argc, char** argv)
{
  MyLogger::Init();
  LOG4CPLUS_ERROR(logger, "Main");
}
```

Log4cplus is inspired by Log4j, should an instance of the Logger be stored in each class, which propagates headers in the case of a library ?

```c++
// MyLibrary

#include <log4cplus/logger.h> // include for all user of my library

/*
namespace log4cplus {
    class Logger;
}
*/

class MyLibrary {
public:
    MyLibrary();

    ~MyLibrary() = default;

    void do_stuff();

private:
    const log4cplus::Logger logger;
//    const log4cplus::Logger& logger ;  ?? is forward class valid with getInstance pattern 
};

// MyLibrary.cpp

#include "MyLibrary.hpp"
#include <log4cplus/loggingmacros.h>

MyLibrary::MyLibrary() : logger(log4cplus::Logger::getInstance("MyLibrary")) {}

void MyLibrary::do_stuff() {
    LOG4CPLUS_INFO(logger, "My library is doing stuff");
}
``` 

Is it possible to use a forward class to avoid propagating the includes of the logger?
I have a lot of questions about the best practices for using log4cplus and the examples don't clarify its real-life usage for me.

**Comment by wilx** [comment, marked as answer]
Having a `Logger` as global/static variable works until process and/or DLL exit time. Then you run into [_static initialization order fiasco_](https://en.cppreference.com/w/cpp/language/siof). If you are concerned about exposing log4cplus classes as implementation details in you headers, you should be able to hide it behind [_pImpl_](https://en.cppreference.com/w/cpp/language/pimpl) pattern. Or you can just rely on retrieving the logger by name every time you need to use it, assuming the performance hit incurred is acceptable.

**Reply by JulienChaos** [reply]
Thank you for this response. I understand the risk of using a static variable. I have decided to store the logger in each class as a member variable, plus use pimpl for the API. (Thanks for the reminder)
